lib/Transforms/Scalar/DxilRemoveUnstructuredLoopExits.cpp

```diff
@@ -297,21 +307,25 @@ static void SkipBlockWithBranch(BasicBlock *bb, Value *cond, Loop *L, LoopInfo *
   bb->getTerminator()->eraseFromParent();
   BranchInst::Create(end, body, cond, bb);
 
+  SmallVector<User *, 4> users;
   for (Instruction &inst : *body) {
     PHINode *phi = nullptr;
 
-    for (User *user : inst.users()) {
-      Instruction *user_inst = dyn_cast<Instruction>(user);
-      if (!user_inst)
-        continue;
+    // Put the users in a list, since we're modifying the user list as we go
+    users.clear();
+    for (User *user : inst.users())
+      users.push_back(user);
```

Comment:
I think you could also use the iterator_range constructor of SmallVector here to the same effect and slightly simpler code.

```c++
SmallVector<User *,4> users(inst.users());
```

However, I think the more common approach in this codebase to this problem is something like this replacing the previous loop:

```c++
for (auto I = inst.users().begin(), E = inst.users().end(); I != E; ++I) {
  Instruction *user_inst = dyn_cast<Instruction>(*I);
...
```

Which I think would have the same effect.

It's not terribly important. What's here is fine. I just wanted to point out alternatives you might consider.

Reply:
Doh! Thanks for pointing it out.